pkg/resources/kafka/kafka.go

```diff
@@ -1223,3 +1227,53 @@ func generateServicePortForEListeners(listeners []v1beta1.ExternalListenerConfig
 	}
 	return usedPorts
 }
+
+func (r *Reconciler) reconcileHeadlessService(log logr.Logger) error {
+	if r.KafkaCluster.Spec.HeadlessServiceEnabled {
+		// Delete kafka all broker service
+		allBrokerServiceName := fmt.Sprintf(kafka.AllBrokerServiceTemplate, r.KafkaCluster.Name)
+		err := r.Client.Delete(context.TODO(), &corev1.Service{ObjectMeta: templates.ObjectMeta(allBrokerServiceName, kafka.LabelsForKafka(r.KafkaCluster.Name), r.KafkaCluster)})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				log.V(1).Info("all broker service not found. Continue")
+			} else {
+				return errors.WrapIfWithDetails(err, "could not delete all broker service")
+			}
+		} else {
+			log.V(1).Info("all broker service deleted", "service name", allBrokerServiceName)
+		}
+		// Delete kafka broker service
+		for _, broker := range r.KafkaCluster.Spec.Brokers {
+			serviceName := fmt.Sprintf("%s-%d", r.KafkaCluster.Name, broker.Id)
+			err = r.Client.Delete(context.TODO(), &corev1.Service{ObjectMeta: templates.ObjectMeta(serviceName,
+				util.MergeLabels(
+					kafka.LabelsForKafka(r.KafkaCluster.Name),
+					map[string]string{"brokerId": fmt.Sprintf("%d", broker.Id)},
+				),
+				r.KafkaCluster)})
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					log.V(1).Info("broker service not found. Continue")
+				} else {
+					return errors.WrapIfWithDetails(err, "could not delete broker service", "id", broker.Id)
```

Comment:
Instead of returning immediately on error while deleting the Kafka broker service one by one, it might be more reasonable to cache the error and move on to deleting the next Kafka broker service in the slice. And we will return the error (if any) after finishing processing all the Kafka broker services. What do you think @stoader ?

Reply:
It is ok for me.

Reply:
I think fast fail is ok here. If it cannot delete a service it means we cannot reach API server properly probably the next deletion would fail too. Return error will make a requeue and it will try the deletion again after some time.